zerver/lib/streams.py

```diff
@@ -791,10 +796,23 @@ def can_add_subscribers_to_streams(
     user_recursive_group_ids = set(
         get_recursive_membership_groups(user_profile).values_list("id", flat=True)
     )
+    if allow_default_streams:
```

Comment:
We have `include_default_realm_subscriptions` parameter which makes sure that the user is subscribed to streams that are default at the time of accepting the invitation. So, maybe can just check permission in the same way whether the stream is default stream or not if stream is passed explicitly in `stream_ids` list.

Reply:
`can_add_subscribers_to_streams` is called by `filter_stream_authorization` which is called by `add_subscriptions_backend`. We don't want to allow subscribing to default streams in that case, which is why we are using the `allow_default_streams` boolean to limit it only to the invite code path.

Reply:
I am basically saying that there is just no need of this code to handle default streams as a special case and remove the complete `if allow_default_streams` block.

So, `add_subscriptions_backend` will not allow subscribing to default streams in case user does not have permission based on the setting.

For the invite code, there are two parameters which decide what streams user will subscribe to - `include_default_realm_subscriptions` and `stream_ids`.

So, we always check the permission based on the setting for all streams passed in `stream_ids` irrespective of whether they are default or not.
And we do not check any permission if `include_realm_default_susbcriptions` is `True` and just allow user to subscribe the invited user.

Reply:
Maybe I'm still mis-interpreting the above message, but we will need that block to allow default streams when only some of the default channels are passed in the stream list and include_realm_default_susbcriptions is not set to true. Maybe [this message](https://chat.zulip.org/#narrow/channel/101-design/topic/can.20subscribe.20other.20users.20permission.20invite/near/2035181) will help clarify the intention here?

We want to allow the user to select the default streams they can invite a user to, whether they have individual permission to invite to those streams or not. Right now `include_realm_default_susbcriptions` only allows all default streams to be subscribed at once and not a subset of them.

Reply:
Yes, I get that and I think that we should only allow users to subscribe to default streams irrespective of permission only if `include_realm_default_susbcriptions` is used. Also, when user accepts the invite, that stream may not be the default stream. Maybe we can discuss this first?

Reply:
I read this conversation again, and think we can allow the user to add to just some of the default channels since the permission to add can also change between sending the invite and user accepting the invite, so channel not being default should also be fine.

Reply:
I don't remember if I commented on CZO somewhere, but yeah, I think adding invited users to any subset of the channels that are default when the invite is being created should be allowed.